packages/@react-stately/combobox/src/useComboBoxState.ts

```diff
@@ -174,8 +174,33 @@ export function useComboBoxState<T extends object>(props: ComboBoxStateProps<T>)
       lastValue.current = inputValue;
     }
 
+    // Update the inputValue if the selected item's text differs from the current input value but only if items are controlled.
+    // This is to handle cases where a selectedKey is specified but the items aren't available (async loading), the selected item's text value updates.
+    // Only reset if the user isn't currently within the field so we don't erroneously modify user input.
+    // Trigger only if a selectedKey exists so that allowCustomValue=true user values don't get reset.
+    if (!isFocused && props.items !== lastItems.current && selectedKey != null) {
```

Comment:
Tried a couple different conditions (doing it only after initialRender, comparing inputValue and lastValue to only update when the user hasn't updated text, etc), settled on only resetting/updating the input text when `items` updates if the combobox isn't focused since IMO it would be a strange user experience to have the combobox input value change without your direct interaction

Reply:
What about with children rather than the `items` prop? Maybe we could use the collection for this instead? Do we actually need to check if `props.items !== lastItems.current` or is the below check of whether the selected key's text change sufficient?

Reply:
I had initially tried to omit the `props.items` comparison, but it broke old behavior such as:
- if a user specifies a `inputValue` and `defaultSelectedKey` and they resolved to different values, we preserve the `inputValue` (the selected key text check would catch that these are different and thus coerce the inputValue to match the defaultSelectKey text, attempting to bypass this via a isInitialRender check failed cuz ComboBox actually renders twice initially due to a setMenuWidth that happens in the component)
- if a user specifies a miss matched `defaultInputValue` and `selectedKey`, the same issue as above happens

Thus I resorted to the `props.items !== lastItems.current` check since it would no longer break the above cases if the user had a `defaultItems` ComboBox. For the controlled `items` case, the user can still have mismatched input value and selected keys initially, but only until `items` is modified.

I didn't consider children/collection (forgot children could change if end user does a `.map` and omitted `items/defaultItems`), I could try using that for a comparator instead

Reply:
I think if someone does a `.map` the collection will change on every render, so you might still have this problem. In the first case you mentioned, I think we'd be ok because we should be ignoring this if `inputValue` is controlled anyway. As for the `defaultInputValue` case... I'm not sure what the correct behavior is there. I would almost expect the selected item value to overwrite it once it loads since usually default props only affect the initial render. WDYT?

Reply:
Hm, shouldn't the collection only get invalidated when `props.items` and `props.children` changes (from useCollection)? I logged out the `lastCollection.current !== collection` in the useEffect and it returns false when I type/interact the `with mapped items` combobox story and returns true when I change the items via clicking the button.

I agree that the `defaultInputValue` should be overwritten after the the items load since it wouldn't be the initial render but its a bit iffy tracking what is considered the "initial render". For our ComboBox, IMO the "initial render" state is after the 2nd render that occurs due to `setMenuWidth` being called [here](https://github.com/adobe/react-spectrum/blob/main/packages/@react-spectrum/combobox/src/ComboBox.tsx#L124-L137) but other implementations of ComboBox that use `useComboBoxState` may have their "initial render" actually be the first render that happens. Not sure that made any sense heh.

Reply:
When you do an array map, it returns a new array with entirely new JSX elements, so the problem is that the collection won't match even if it didn't change, e.g. the component was re-rendered for some other reason. That's why we have the `items` prop in the first place (so we can cache the JSX elements). I think we cannot rely on object identity for this.

> For our ComboBox, IMO the "initial render" state is after the 2nd render

But that shouldn't be an issue right, because we'd only reset the input value when the selectedKey becomes available in the collection. After that second initial render it still wouldn't have loaded yet so nothing would happen.

Reply:
> When you do an array map, it returns a new array with entirely new JSX elements...
Ah, bleh I see. I'll try and think of an alternative then (I'd like to just use the key's text vs the current input text but the defaultInputValue case is annoying)  

> But that shouldn't be an issue right, because we'd only reset the input value when the selectedKey...
It's an issue for a `defaultItems`/static ComboBox since the collection is available immediately. In this case it overwrites what the user sets as the `defaultInputValue` since the 2nd render from `setMenuWidth` will update the inputValue with the `selectedKey` text.

Reply:
try changing the condition to be if the selectedKey's text changes: store ref for lastSelectedKeyText, on each render check if selectedKey.textValue != lastSelectedKeyText and reset if true